_config/routes.yml

```diff
@@ -11,7 +11,7 @@ After:
 ---
 SilverStripe\Control\Director:
   rules:
-    'Security//$Action/$ID/$OtherID': SilverStripe\Security\Security
+    'Security//$Action/$ID/$OtherID': '%$RateLimitedSecurityController'
```

Comment:
Should we allow adding middleware to route definitions directly? It's not really composable this way. Might also be worth pointing out in the upgrade guide, since I've seen devs overwriting their `Security` director rules to direct to custom controllers.

Reply:
We could nest this as another composable level: e.g.


```yaml
SilverStripe\Control\Director:
  rules:
    'Security//$Action/$ID/$OtherID': '%$SecurityController'
---
Name: securitystuff
---
SilverStripe\Core\Injector\Injector:
  SecurityController: '%$RateLimitedSecurityController'
```

Reply:
However you configure the class use (route or injector), it's still a different class that you might accidentally overwrite as a dev without realising what you're doing. Our YAML-based dependency injection is powerful, but not very transparent - there's no good way to list the "final" injection rules and determine if you've overwritten another rule, right?

I don't see this as a controller wrapper, but as a route specific middleware, which is a feature that doesn't exist in core at the moment. Sorry, but I think we should defer this implementation until it does.

Reply:
>I don't see this as a controller wrapper, but as a route specific middleware, which is a feature that doesn't exist in core at the moment. Sorry, but I think we should defer this implementation until it does.

This is literally in the docs as how to implement route-specific middleware. I don't see that it makes sense to block this PR because the current way to implement route-specific middleware isn't deemed to be suitable.

We can address this in 4.x branch once 4.0 is out if we are concerned about the implementation of route-specific middlewares, but to block a fairly major security enhancement (which will make SS more secure out of the box) on these grounds seems to go against what's best for the framework and our users.